## Researching Commands: grep
For this lab report, we are going to research `grep` and four of its command-line options. In order to see a quick rundown of the various possible options, I will type `grep --help` into the terminal (this won't count as one of the four). For my own reference, I'll paste the list below:
```
Usage: grep [OPTION]... PATTERN [FILE]...
Search for PATTERN in each FILE or standard input.
PATTERN is, by default, a basic regular expression (BRE).
Example: grep -i 'hello world' menu.h main.c

Regexp selection and interpretation:
  -E, --extended-regexp     PATTERN is an extended regular expression (ERE)
  -F, --fixed-strings       PATTERN is a set of newline-separated strings
  -G, --basic-regexp        PATTERN is a basic regular expression (BRE)
  -P, --perl-regexp         PATTERN is a Perl regular expression
  -e, --regexp=PATTERN      use PATTERN for matching
  -f, --file=FILE           obtain PATTERN from FILE
  -i, --ignore-case         ignore case distinctions
  -w, --word-regexp         force PATTERN to match only whole words
  -x, --line-regexp         force PATTERN to match only whole lines
  -z, --null-data           a data line ends in 0 byte, not newline

Miscellaneous:
  -s, --no-messages         suppress error messages
  -v, --invert-match        select non-matching lines
  -V, --version             display version information and exit
      --help                display this help text and exit

Output control:
  -m, --max-count=NUM       stop after NUM matches
  -b, --byte-offset         print the byte offset with output lines
  -n, --line-number         print line number with output lines
      --line-buffered       flush output on every line
  -H, --with-filename       print the file name for each match
  -h, --no-filename         suppress the file name prefix on output
      --label=LABEL         use LABEL as the standard input file name prefix
  -o, --only-matching       show only the part of a line matching PATTERN
  -q, --quiet, --silent     suppress all normal output
      --binary-files=TYPE   assume that binary files are TYPE;
                            TYPE is 'binary', 'text', or 'without-match'
  -a, --text                equivalent to --binary-files=text
  -I                        equivalent to --binary-files=without-match
  -d, --directories=ACTION  how to handle directories;
                            ACTION is 'read', 'recurse', or 'skip'
  -D, --devices=ACTION      how to handle devices, FIFOs and sockets;
                            ACTION is 'read' or 'skip'
  -r, --recursive           like --directories=recurse
  -R, --dereference-recursive  likewise, but follow all symlinks

'egrep' means 'grep -E'.  'fgrep' means 'grep -F'.
Direct invocation as either 'egrep' or 'fgrep' is deprecated.
When FILE is -, read standard input.  With no FILE, read . if a command-line
-r is given, - otherwise.  If fewer than two FILEs are given, assume -h.
Exit status is 0 if any line is selected, 1 otherwise;
if any error occurs and -q is not given, the exit status is 2.
```

## Using grep -f

This command-line option attempts to search files based on a given pattern from a file and returns the lines that adhere to that pattern. It follows the syntax:\
`grep -f patterns_file search_file(s)`

### Example 1
```
$ grep -f plos/test-pattern.txt plos/journal.pbio.0020001.txt
        States and Canada by the year 2000 (Figure 2). Although the cost of research is undoubtedly
        Although there are outstanding scientific researchers in the developing world who
        Although the evidence presented here demonstrates that there is a long way to go before
        target the journals that have the greatest impact. Although there may still be a long road
```

In the example above, I created a new text file inside of the plos folder to learn the functionality of `grep -f`. I typed "Although the" into the file and named it `test-pattern.txt` and then provided one of the journal documents as input to be searched in. The four lines returned are all lines that contained the pattern "Although the". It's immediately apparent why this is useful; it provides a function similar to Ctrl+F, which allows us to find specific text patterns instantly regardless of a document's size.

### Example 2
```
$ grep -f plos/test-pattern.txt plos/journal.pbio.0020001.txt plos/journal.pbio.0020401.txt
plos/journal.pbio.0020001.txt:        States and Canada by the year 2000 (Figure 2). Although the cost of research is undoubtedly
plos/journal.pbio.0020001.txt:        Although there are outstanding scientific researchers in the developing world who
plos/journal.pbio.0020001.txt:        Although the evidence presented here demonstrates that there is a long way to go before
plos/journal.pbio.0020001.txt:        target the journals that have the greatest impact. Although there may still be a long road
```

The above example is similar, but different! When you add a third file into the command, it will search through the additional file as well. This means that any file after the first (which is the pattern file) will be searched. In the output, the filepath on the left seems to indicate which file the lines are coming from. So why didn't the second search file have any lines? It does contain the phrase "Although the", but it is written as "although the". The pattern is case-sensitive!

While I found `grep -f` in the list provided by the terminal, I discovered how to use it by asking ChatGPT with the following prompt:\
"In Bash, what's an example of how I can use `grep -f`?"\
It responded with a simple definition of the command as a way to search for multiple patterns stored in a file and then gave me an example, with text files like so:
```
patterns.txt:            fruits.txt:
apple                    apple
banana                   mango
cherry                   banana
                         orange
                         cherry
```
After writing the command `grep -f patterns.txt fruits.txt`, the output produced was:
```
apple
banana
cherry
```
This means that the lines specified in pattern.txt were successfully found in fruits.txt, as they were all returned. In order to verify my understanding, I made text files on my `@ieng6` account with the same names and contents, and then ran the same command. The terminal produced the same output, so I could be certain that both ChatGPT's response was correct and my understanding was solidified.

## Using grep -e

This command-line option attempts to search files based on a given pattern directly on the command-line and returns the lines that adhere to that pattern. It follows the syntax:\
`grep -e "pattern" search_file(s)`

### Example 1
```
$ grep -e "R26R" biomed/*.txt
biomed/1471-213X-1-4.txt:        lacZ reporter strain (R26R) was
biomed/1471-213X-1-4.txt:        R26R allele terminates transcription
biomed/1471-213X-1-4.txt:        colonies were analyzed for each construct. Three R26R-EYFP
biomed/1471-213X-1-4.txt:        and two R26R-ECFP colonies carried the targeted allele, as
biomed/1471-213X-1-4.txt:        detected in the R26R-EYFP or R26R-ECFP mice, as expected
biomed/1471-213X-1-4.txt:        The specificity of the R26R-EYFP and ECFP lines was
biomed/1471-213X-1-4.txt:        25]. Figure 3Ashows an experiment in which the R26R-YFP
biomed/1471-213X-1-4.txt:        in the doubly transgenic offspring. Thus R26R-YFP mice
biomed/1471-213X-1-4.txt:        R26R lacZ allele [ 4], resulting in a
biomed/1471-213X-1-4.txt:        R26R-YFP mice were crossed with a strain of mice in which
biomed/1471-213X-1-4.txt:        R26R lacZ allele [ 4], resulting in
biomed/1471-213X-1-4.txt:        their similarity to the well characterized R26R-lacZ
biomed/1471-213X-1-4.txt:          To make the R26R-YFP targeting construct, the
biomed/1471-213X-1-4.txt:          To make the R26R-CFP targeting construct pECFP was
biomed/1471-213X-1-4.txt:          electroporations with each of the two constructs R26R-YFP
biomed/1471-213X-1-4.txt:          and R26R-CFP. Genomic Southern blot hybridization was
biomed/1471-213X-1-4.txt:          R26R-lacZ reporter mice were a gift of Dr. Philippe
```

In the above example, I picked a cool looking word as the pattern and had it search through the entire biomed folder for it. It turns out that only one file contained the pattern. In retrospect, this command is closer to the function of Ctrl+F than `grep -f`. `grep -f` might find better use when working with file redirection and scripts to search things in a more automated manner.

### Example 2
```
$ grep -e "step-wise" biomed/*.txt
biomed/1471-2261-1-6.txt:          adherence to step-wise logistic regression to identify
biomed/1471-2261-1-6.txt:        not chosen by the step-wise procedure. Age, gender, and
biomed/1471-2261-1-6.txt:        over-reporting in both the crude and step-wise logistic
biomed/1471-2288-2-4.txt:        results. Overfitting is often associated with step-wise
biomed/1472-6904-1-2.txt:            step-wise ANCOVA) fail to sustain significance. One of
biomed/1475-925X-2-10.txt:            step-wise increased. The process terminates when all
biomed/gb-2001-2-8-research0031.txt:        the possible clusters of genes in a step-wise regression
biomed/gb-2002-3-12-research0072.txt:          We continued the step-wise model enlargement and
biomed/gb-2002-3-12-research0072.txt:          maximization in a step-wise manner and so, as in any
```

For this example, I wanted to pick a term that was in multiple files to see how the output appears when using star for the filepath. It functions as I expected; the different files are listed, all from the same directory. Something I noticed is that the indentation is the same size between the filepath and printed line, which could potentially make it difficult to read down the column. The alternative, however, would cause the entire list to be pushed far away if one particular filename was extremely long.

I again used ChatGPT to get a better understanding. To discover how `grep -e` is different from `grep -f`, I asked:\
"In Bash, how does `grep -e` differ from `grep -f`?"

## Using grep -r

This command-line option recursively searches an entire directory and its subdirectories for a pattern. It follows the syntax:\
`grep -r "pattern" directory`

### Example 1
```
$ grep -r "R26R" biomed
biomed/1471-213X-1-4.txt:        lacZ reporter strain (R26R) was
biomed/1471-213X-1-4.txt:        R26R allele terminates transcription
biomed/1471-213X-1-4.txt:        colonies were analyzed for each construct. Three R26R-EYFP
biomed/1471-213X-1-4.txt:        and two R26R-ECFP colonies carried the targeted allele, as
biomed/1471-213X-1-4.txt:        detected in the R26R-EYFP or R26R-ECFP mice, as expected
biomed/1471-213X-1-4.txt:        The specificity of the R26R-EYFP and ECFP lines was
biomed/1471-213X-1-4.txt:        25]. Figure 3Ashows an experiment in which the R26R-YFP
biomed/1471-213X-1-4.txt:        in the doubly transgenic offspring. Thus R26R-YFP mice
biomed/1471-213X-1-4.txt:        R26R lacZ allele [ 4], resulting in a
biomed/1471-213X-1-4.txt:        R26R-YFP mice were crossed with a strain of mice in which
biomed/1471-213X-1-4.txt:        R26R lacZ allele [ 4], resulting in
biomed/1471-213X-1-4.txt:        their similarity to the well characterized R26R-lacZ
biomed/1471-213X-1-4.txt:          To make the R26R-YFP targeting construct, the
biomed/1471-213X-1-4.txt:          To make the R26R-CFP targeting construct pECFP was
biomed/1471-213X-1-4.txt:          electroporations with each of the two constructs R26R-YFP
biomed/1471-213X-1-4.txt:          and R26R-CFP. Genomic Southern blot hybridization was
biomed/1471-213X-1-4.txt:          R26R-lacZ reporter mice were a gift of Dr. Philippe
```

For the above example, I once again used the cool looking word to see if it would work in grep -r, and it did. Instead of taking all of the .txt files in biomed as input like grep -e, this option takes the directory itself as the input. I imagine this is useful when you are unaware of the file that contains the pattern in an absurdly long list of files. I'm unsure just how different it is from grep -e, but it is definitely a more accessible option for those unfamiliar with path syntax.

### Example 2
```
$ grep -r "Kelly Greene" government
government/Media/A_Perk_of_Age.txt:By Kelly Greene
```

For this example, I wanted to see the "and subdirectories" part of the function in action. "Kelly Greene" can only be found in one file, which is inside a subdirectory named Media. It was able to do so with the same speed as the other command, which surprised me. This seems to be the main benefit of using grep -r, as none of the other commands so far can search such a broad area with a simple directory name.

Don't worry, I didn't **just** use ChatGPT this time. I also used what seems to be the official website to learn about grep -r:\
https://www.gnu.org/software/grep/manual/grep.html

## Using grep -c

This command-line option searches files based on a given pattern and returns the count of the number of lines that contain the pattern. It follows the syntax:\
`grep -c "pattern" search_file(s)`

### Example 1
```
$ grep -c "R26R" biomed/1471-213X-1-4.txt
17
```

In the above example, I used the cool looking word one last time to see what the output looks like with grep -c. It returned the number 17, which checks out with the number of lines that contained the term R26R. I imagine this can be useful in cases where the lines themselves don't provide useful information and you simply want to know how often it was said.

### Example 2
```
$ grep -c "woowaejdskglfsdhghsf" biomed/*.txt
biomed/1468-6708-3-1.txt:0
biomed/1468-6708-3-10.txt:0
biomed/1468-6708-3-3.txt:0
biomed/1468-6708-3-4.txt:0
biomed/1468-6708-3-7.txt:0
biomed/1471-2091-2-10.txt:0
biomed/1471-2091-2-11.txt:0
biomed/1471-2091-2-12.txt:0
biomed/1471-2091-2-13.txt:0
biomed/1471-2091-2-16.txt:0
biomed/1471-2091-2-5.txt:0
...
```

In this example, I wanted to use star again to see if grep -c could search the entire directory and print the number of lines that matched for each file. I noticed that it was not working, so I typed gibberish as the pattern and it still returned every file in the directory. I am unsure if I did something wrong or grep -c simply can't be used in this manner.

As for the discovery, I used the GNU website once more to read about grep -c.
